platform/darwin/test/MGLFilterTests.m

```diff
@@ -0,0 +1,106 @@
+#import "MGLStyleLayerTests.h"
+
+@interface MGLFilterTests : MGLStyleLayerTests {
+    MGLGeoJSONSource *source;
+    MGLLineStyleLayer *layer;
+}
+@end
+
+@implementation MGLFilterTests
+
+- (void)setUp
+{
+    [super setUp];
+    NSString *filePath = [[NSBundle bundleForClass:self.class] pathForResource:@"amsterdam" ofType:@"geojson"];
+    NSURL *url = [NSURL fileURLWithPath:filePath];
+    NSData *geoJSONData = [NSData dataWithContentsOfURL:url];
+    source = [[MGLGeoJSONSource alloc] initWithSourceIdentifier:@"test-source" geoJSONData:geoJSONData];
+    [self.mapView.style addSource:source];
+    layer = [[MGLLineStyleLayer alloc] initWithLayerIdentifier:@"test-layer" sourceIdentifier:@"test-source"];
+}
+
+- (void)tearDown
+{
+    [self.mapView.style removeLayer:layer];
+    [self.mapView.style removeSource:source];
+}
+
+- (NSArray<NSPredicate *> *)predicates
+{
+    NSPredicate *equalPredicate = [NSPredicate predicateWithFormat:@"type == 'neighbourhood'"];
+    NSPredicate *notEqualPredicate = [NSPredicate predicateWithFormat:@"type != 'park'"];
+    NSPredicate *greaterThanPredicate = [NSPredicate predicateWithFormat:@"%K > %@", @"stroke-width", @2.1];
+    NSPredicate *greaterThanOrEqualToPredicate = [NSPredicate predicateWithFormat:@"%K >= %@", @"stroke-width", @2.1];
+    NSPredicate *lessThanOrEqualToPredicate = [NSPredicate predicateWithFormat:@"%K <= %@", @"stroke-width", @2.1];
+    NSPredicate *lessThanPredicate = [NSPredicate predicateWithFormat:@"%K < %@", @"stroke-width", @2.1];
+    NSPredicate *inPredicate = [NSPredicate predicateWithFormat:@"type IN %@", @[@"park", @"neighbourhood"]];
+    NSPredicate *notInPredicate = [NSPredicate predicateWithFormat:@"NOT (type IN %@)", @[@"park", @"neighbourhood"]];
+    NSPredicate *inNotInPredicate = [NSPredicate predicateWithFormat:@"type IN %@ AND NOT (type IN %@)", @[@"park"], @[@"neighbourhood", @"test"]];
+    NSPredicate *typePredicate = [NSPredicate predicateWithFormat:@"%K == %@", @"$type", @"Feature"];
+    NSPredicate *idPredicate = [NSPredicate predicateWithFormat:@"%K == %@", @"$id", @"1234123"];
+    NSPredicate *specialCharsPredicate = [NSPredicate predicateWithFormat:@"%K == %@", @"ty-’pè", @"sŒm-ethįng"];
+    return @[equalPredicate,
+             notEqualPredicate,
+             greaterThanPredicate,
+             greaterThanOrEqualToPredicate,
+             lessThanOrEqualToPredicate,
+             lessThanPredicate,
+             inPredicate,
+             notInPredicate,
+             inNotInPredicate,
+             typePredicate,
+             idPredicate,
+             specialCharsPredicate];
+}
+
+- (void)testAllPredicates
+{
+    for (NSPredicate *predicate in self.predicates) {
+        layer.predicate = predicate;
+        XCTAssertEqualObjects(layer.predicate, predicate);
```

Comment:
This ensures we can round-trip the predicates, which is a good start. However, the tests don’t verify that for instance `specialCharsPredicate` resolves to the right `mbgl::style::Filter`. What if the special characters get mangled during the conversion but get unmangled during the conversion back to NSPredicate?